sw/device/lib/dif/README.md

```diff
@@ -12,3 +12,56 @@ libraries.
 There is one DIF library per hardware IP, and each one contains the DIFs
 required to actuate all of the specification-required functionality of the
 hardware they are written for.
+
+## DIF Style Guide
+
+DIFs are very low-level software, so they have a more rigorous coding style than
+other parts of the codebase.
+
+DIFs should follow the [OpenTitan C/C++ style
+guide](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) where it does not
+contradict with the guidelines below.
+
+The guidelines below apply to writing DIFs, and code should be written in a
+similar style to the existing DIF libraries in this directory.
+
+### DIF Library Guidance
+
+* DIF libraries must be written in C.
+* DIF libraries can only depend on the headers and libraries in the
+  `sw/device/lib/base` directory.
+* DIF libraries must not depend on other DIF libraries. Exercising DIF
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can either error or produce a value, it should return an enum return
+  code, and use an out-parameter for returning the produced value.
```

Comment:
This does not address what an infaillable DIF function should return (which is a potential thing that could exist). I'm in favor of
```
typedef enum dif_infaillable {
  kDifInfaillableOk,
} dif_infaillable_t;
```

Reply:
Infallible DIFs should return a value (if there is one to return), or `void` if there is not. Infallible DIFs should not use out-parameters if they return `void`. I don't see any value in returning `dif_infaillable_t` where `void` would suffice.

Reply:
> Infallible DIFs should return a value (if there is one to return), or `void` if there is not. Infallible DIFs should not use out-parameters if they return `void`. I don't see any value in returning `dif_infaillable_t` where `void` would suffice.

Strongly agree.  It behooves callers to check the return values of functions.  If a function returns `dif_infalliable_t` I would expect them to check its value.  The cycle of "returning a meaningless value and then checking the meaningless value" strikes me as a high amount of overhead for not writing `void`.